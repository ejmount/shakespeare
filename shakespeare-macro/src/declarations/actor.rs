use itertools::Itertools;
use quote::ToTokens;
use syn::{parse_quote, Error, ImplItem, ItemImpl, ItemMod, Type, Visibility};

use super::performance::{PerformanceAttribute, PerformanceDeclaration};
use super::role::RoleDecl;
use crate::data::{ActorName, DataItem};
use crate::macros::{fallible_quote, filter_unwrap};

pub struct ActorDecl {
	pub actor_name:   ActorName,
	pub actor_vis:    Visibility,
	pub data_item:    DataItem,
	pub performances: Vec<PerformanceDeclaration>,
	pub roles:        Vec<RoleDecl>,
	pub others:       Vec<ItemImpl>,
}

impl ActorDecl {
	pub fn new(module: ItemMod) -> Result<ActorDecl, Error> {
		let inline_err = Err(Error::new_spanned(
			&module,
			"Actor declaration cannot be empty",
		));
		let missing_err = Err(Error::new_spanned(
			&module,
			"Actor declaration must contain one struct, enum or union",
		));
		let Some((_, items)) = module.content else {
			return inline_err;
		};

		let actor_vis = module.vis;

		let data_item = match items.iter().flat_map(DataItem::try_from).at_most_one() {
			Ok(Some(item)) => item,
			Ok(None) => return missing_err,
			Err(extras) => {
				let errors = extras.map(|d| {
					Error::new_spanned(d, "Only one data item allowed in actor declaration")
				});
				return Err(errors.reduce(combine_errors).unwrap());
			}
		};

		let actor_name = module.ident;
		let actor_name = fallible_quote! { #actor_name }?;

		let mut performances = vec![];
		let mut roles = vec![];
		let mut others = vec![];

		for item in items {
			if let syn::Item::Impl(imp) = item {
				if let Some(attr) = get_performance_tag(&imp) {
					let arg: PerformanceAttribute = attr.parse_args()?;
					let canonical = arg.canonical;

					let role_name = imp.trait_.clone().unwrap().1;

					if canonical {
						let signatures = filter_unwrap!(imp.items.iter(), ImplItem::Fn)
							.map(|f| &f.sig)
							.cloned();

						let role = RoleDecl::new(role_name.clone(), parse_quote! {pub}, signatures);
						roles.push(role);
					}

					performances.push(PerformanceDeclaration::new(role_name, imp)?);
				} else if let Type::Path(p) = &*imp.self_ty {
					if p.path == data_item.name().0 {
						others.push(imp);
					} else {
						panic!(
							"Path: {}, ident: {:?}",
							p.path.to_token_stream(),
							data_item.name(),
						);
					}
				} else {
					continue;
				}
			}
		}
		assert!(!performances.is_empty(), "Empty perfs"); // Because [SpawningFunction] falls over otherwise

		Ok(ActorDecl {
			actor_name,
			actor_vis,
			data_item,
			performances,
			roles,
			others,
		})
	}
}

fn get_performance_tag(imp: &syn::ItemImpl) -> Option<&syn::Attribute> {
	imp.attrs.iter().find(|attr| {
		attr.path()
			.segments
			.last()
			.is_some_and(|ps| ps.ident == "performance")
	})
}

fn combine_errors(mut one: Error, another: Error) -> Error {
	one.combine(another);
	one
}
