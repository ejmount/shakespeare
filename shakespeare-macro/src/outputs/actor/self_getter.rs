use quote::ToTokens;
use syn::{Item, Result};

use crate::data::{ActorName, DataName};
use crate::macros::fallible_quote;

#[derive(Debug)]
pub(crate) struct SelfGetter {
	statik_item:  Item,
	data_getter:  Item,
	actor_getter: Item,
}

impl SelfGetter {
	pub(crate) fn new(actor_name: &ActorName, data_name: &DataName) -> Result<SelfGetter> {
		let getter_item = actor_name.get_static_item_name();

		let statik_item: Item = fallible_quote! {
			::shakespeare::tokio_export::task_local! {
				#[doc(hidden)]
				static #getter_item: ::std::sync::Arc<#actor_name>;
			}
		}?;

		let data_getter: Item = fallible_quote! {
			impl #data_name {
				/// Gets a reference to the running actor
				///
				/// # Panics
				/// Will panic if called from outside a performance of an actor of the appropriate type. (However, which instance it's called on doesn't matter.)
				///
				#[allow(dead_code)]
				pub fn get_shell(&self) -> ::std::sync::Arc<#actor_name> {
					#getter_item.with(Clone::clone)
				}
			}
		}?;

		let actor_getter: Item = fallible_quote! {
			impl #actor_name {
				#[doc(hidden)]
				// Used internally for creating Envelopes
				pub fn get_shell(&self) -> ::std::sync::Arc<#actor_name> {
					self.this.upgrade().expect("Dead actor?")
				}
			}
		}?;

		Ok(SelfGetter {
			statik_item,
			data_getter,
			actor_getter,
		})
	}
}
impl ToTokens for SelfGetter {
	fn to_tokens(&self, tokens: &mut proc_macro2::TokenStream) {
		self.statik_item.to_tokens(tokens);
		self.data_getter.to_tokens(tokens);
		self.actor_getter.to_tokens(tokens);
	}
}
