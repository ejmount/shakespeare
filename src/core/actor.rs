use std::fmt::Debug;
use std::sync::Arc;
use std::task::{Context, Poll};

use futures::Future;
use tokio::task::JoinHandle;

#[non_exhaustive]
/// Indicates whether an actor closed down successfully and any output value produced
pub enum Outcome<A: Shell> {
	#[doc(hidden)]
	/// The actor was explicitly aborted at the task level.
	/// It shouldn't be possible to encounter this is in practice,
	Aborted(tokio::task::JoinError),
	/// The actor exited successfully either as all of its strong references dropped or by explicit shutdown.
	Exit(A::ExitType),
	/// The actor panicked executing one of its message handlers.
	Panic(A::PanicType),
}

impl<A: Shell> Debug for Outcome<A> {
	fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
		match self {
			Outcome::Aborted(_) => f.write_str("ActorOutcome::Aborted"),
			Outcome::Exit(_) => f.write_str("ActorOutcome::Exit"),
			Outcome::Panic(_) => f.write_str("ActorOutcome::Panic"),
		}
	}
}

impl<A: Shell> PartialEq for Outcome<A>
where
	A::ExitType: PartialEq,
	A::PanicType: PartialEq,
{
	fn eq(&self, other: &Self) -> bool {
		use Outcome::{Exit, Panic};
		match (self, other) {
			(Exit(a), Exit(b)) => a == b,
			(Panic(a), Panic(b)) => a == b,
			_ => false,
		}
	}
}

impl<A: Shell> Eq for Outcome<A>
where
	A::ExitType: Eq,
	A::PanicType: Eq,
{
}

/// A handle for the actor's exit value.
///
/// Serves the same role as [`std::thread::JoinHandle`], but for an actor. Can be awaited to receive the actor's output value.
/// As with `JoinHandle`, dropping this value without awaiting it detaches the actor task and makes any output value from the actor inaccessible.
pub struct Handle<A: Shell>(JoinHandle<Result<A::ExitType, A::PanicType>>);

impl<A: Shell> Handle<A> {
	fn new(val: JoinHandle<Result<A::ExitType, A::PanicType>>) -> Handle<A> {
		Handle(val)
	}
}

impl<A: Shell> Debug for Handle<A> {
	fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
		f.write_str("ActorHandle")
	}
}

impl<A: Shell> Future for Handle<A> {
	type Output = Outcome<A>;

	fn poll(self: std::pin::Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
		let handle = &mut self.get_mut().0;
		tokio::pin!(handle);
		match handle.poll(cx) {
			Poll::Pending => Poll::Pending,
			Poll::Ready(result) => match result {
				Ok(Ok(e)) => Outcome::Exit(e),
				Ok(Err(f)) => Outcome::Panic(f),
				Err(e) => Outcome::Aborted(e),
			}
			.into(),
		}
	}
}

/// A trait indicating the message-handling external interface of an actor. Implementations will be generated by macros - you should not implement this manually.
pub trait Shell {
	/// The internal state type within this shell
	type StateType: State<ShellType = Self>;
	/// The output type yielded by the actor exiting successfully
	type ExitType;
	/// The output type yielded by the actor's panic handler after a message handler panicked
	type PanicType;
}

/// The state type of an actor. This trait is for macro use only.
pub trait State {
	/// The internal state type within this shell
	type ShellType: Shell<StateType = Self>;
}

/// Various values used to interact with an actor that has just been `start()`d.
///
/// The most important value here is the `Arc<A>`, which is the handle needed to send messages to the actor. This is ref-counted, and the actor will begin the process of shutting down when the last handle drops.
///
/// The `Handle<A>` is a future that enables awaiting on the actor's completion and retrieving its output value, either as a result of the last `Arc<A>` dropping or theh actor's message handlers panicking. However, if you do not need this value, the `Handle` can be discarded and the actor will continue running.
#[non_exhaustive]
#[derive(Debug)]
pub struct Spawn<A>
where
	A: Shell,
{
	/// A handle for sending messages to the actor
	pub actor_handle: Arc<A>,
	/// A future for awaiting the actor's completion
	pub join_handle:  Handle<A>,
}

impl<A: Shell> Spawn<A> {
	#[doc(hidden)]
	pub fn new(actor: Arc<A>, handle: JoinHandle<Result<A::ExitType, A::PanicType>>) -> Spawn<A> {
		Spawn {
			actor_handle: actor,
			join_handle:  Handle::new(handle),
		}
	}
}
