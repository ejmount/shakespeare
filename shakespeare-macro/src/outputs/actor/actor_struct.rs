use itertools::Itertools;
use quote::{quote, ToTokens};
use syn::parse::Parser;
use syn::{
	Field, ImplItem, ItemFn, ItemImpl, ItemStruct, Result, ReturnType, Signature, Visibility,
};

use crate::data::{ActorName, DataName, RoleName};
use crate::declarations::{ActorDecl, PerformanceDecl};
use crate::macros::{fallible_quote, map_or_bail};

#[derive(Debug)]
pub(crate) struct ActorStruct {
	strukt:                  ItemStruct,
	sender_method_name_impl: ItemImpl,
	meta_traits:             [ItemImpl; 2],
}

impl ActorStruct {
	pub(crate) fn new(actor: &ActorDecl) -> Result<ActorStruct> {
		let ActorDecl {
			actor_name,
			performances,
			actor_vis,
			panic_handler,
			exit_handler,
			data_item,
			..
		} = actor;

		let fields = map_or_bail!(performances, shell_field_from_performance);

		let strukt = fallible_quote! {
			#actor_vis struct #actor_name {
				this: ::std::sync::Weak<Self>,
				#(#fields),*
			}
		}?;

		let role_names = performances
			.iter()
			.map(PerformanceDecl::get_role_name)
			.collect_vec();

		let sender_method_name_impl = create_inherent_impl(&role_names, actor_vis, actor_name)?;

		let meta_trait = create_meta_trait_impl(
			panic_handler.as_ref(),
			exit_handler.as_ref(),
			actor_name,
			&data_item.name(),
		)?;

		Ok(ActorStruct {
			strukt,
			sender_method_name_impl,
			meta_traits: meta_trait,
		})
	}
}

impl ToTokens for ActorStruct {
	fn to_tokens(&self, tokens: &mut proc_macro2::TokenStream) {
		self.strukt.to_tokens(tokens);
		self.meta_traits.each_ref().map(|f| f.to_tokens(tokens));
		self.sender_method_name_impl.to_tokens(tokens);
	}
}

fn create_meta_trait_impl(
	panic_handler: Option<&ItemFn>,
	exit_handler: Option<&ItemFn>,
	actor_name: &ActorName,
	data_name: &DataName,
) -> Result<[ItemImpl; 2]> {
	let unit_type = fallible_quote!(())?;
	let panic_type = fallible_quote!(std::boxed::Box<dyn std::any::Any + std::marker::Send>)?;

	let panic_return = match panic_handler.as_ref() {
		Some(ItemFn {
			sig: Signature {
				output: ReturnType::Type(_, b),
				..
			},
			..
		}) => &**b,
		Some(_) => &unit_type,
		None => &panic_type,
	};
	let exit_return = if let Some(ItemFn {
		sig: Signature {
			output: ReturnType::Type(_, b),
			..
		},
		..
	}) = exit_handler.as_ref()
	{
		&**b
	} else {
		&unit_type
	};

	let actor_trait = fallible_quote! {
		impl ::shakespeare::ActorShell for #actor_name {
			type StateType = #data_name;
			type ExitType = #exit_return;
			type PanicType = #panic_return;
		}
	}?;

	let state_trait = fallible_quote! {
		impl ::shakespeare::ActorState for #data_name {
			type ShellType = #actor_name;
		}
	}?;

	Ok([actor_trait, state_trait])
}

fn create_inherent_impl(
	role_names: &Vec<&RoleName>,
	actor_vis: &Visibility,
	actor_name: &ActorName,
) -> Result<ItemImpl> {
	let make_methods = |role_name: &&RoleName| -> Result<ImplItem> {
		let field_name = role_name.queue_name();
		let acccessor_name = role_name.sender_method_name();

		fallible_quote! {
			#actor_vis async fn #acccessor_name(&self, payload: ::shakespeare::ReturnEnvelope<dyn #role_name>) -> Result<(), ::shakespeare::Role2SendError<dyn #role_name>>
			{
				self.#field_name.send(payload)
			}
		}
	};

	let methods = map_or_bail!(role_names, make_methods);

	fallible_quote! {
		impl #actor_name {
			#(#methods)*
		}
	}
}

fn shell_field_from_performance(perf: &PerformanceDecl) -> Result<Field> {
	let role_name = &perf.role_name;
	let field_name = role_name.queue_name();

	Field::parse_named
		.parse2(quote! {#field_name : shakespeare::Role2Sender<dyn #role_name> })
		.map_err(|err| {
			syn::parse::Error::new(err.span(),
				format!("Parse failure trying to create actor field: {err} - this is a bug, please file an issue")
			)
		})
}
