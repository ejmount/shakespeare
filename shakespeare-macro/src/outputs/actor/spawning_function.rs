use itertools::{Itertools, izip};
use proc_macro2::TokenStream;
use quote::{ToTokens, format_ident};
use syn::parse::Parser;
use syn::{Expr, Field, Ident, ItemImpl, Result, Stmt};

use crate::data::{ActorName, DataName, RoleName};
use crate::declarations::PerformanceDecl;
use crate::macros::{fallible_quote, map_or_bail};

#[derive(Debug)]
pub(crate) struct SpawningFunction {
	fun: ItemImpl,
}

impl SpawningFunction {
	pub(crate) fn new(
		actor_name: &ActorName,
		data_name: &DataName,
		performances: &[PerformanceDecl],
		panic_name: Option<&Ident>,
		exit_name: Option<&Ident>,
	) -> Result<SpawningFunction> {
		let field_names = performances
			.iter()
			.map(PerformanceDecl::get_role_name)
			.map(RoleName::queue_name)
			.collect_vec();

		let input_field_names = field_names
			.iter()
			.map(|name| format_ident!("{}_input", name))
			.collect_vec();

		let output_field_names = field_names
			.iter()
			.map(|name| format_ident!("{}_output", name))
			.collect_vec();

		let queue_constructions = map_or_bail!(
			izip!(performances, &input_field_names, &output_field_names),
			|(role, inn, out)| -> Result<Stmt> {
				let role_name = &role.role_name;
				fallible_quote! { let (#inn, mut #out) = <dyn #role_name as ::shakespeare::Role>::Channel::new(Default::default()); }
			}
		);

		let actor_fields =
			map_or_bail!(
				izip!(performances, &input_field_names),
				|(role, input)| -> Result<Field> {
					let field_name = role.role_name.queue_name();
					Field::parse_named.parse2(fallible_quote! {#field_name : #input}?)
				}
			);

		assert!(!performances.is_empty());
		assert!(!output_field_names.is_empty());

		let select_branches = map_or_bail!(
			izip!(performances, &output_field_names),
			|(perf, output)| -> Result<TokenStream> {
				let fn_name = perf.role_name.method_name();
				fallible_quote! { Some(msg) = #output.recv(), if !(#output.is_empty() && !context.sustains()) => {
					timeout_sleep.as_mut().reset(Instant::now() + IDLE_TIMEOUT);
					state.#fn_name(&mut context, msg).await
				} }
			}
		);
		assert!(!select_branches.is_empty());

		let constructor: Expr = fallible_quote! {
			#actor_name {
				this: weak.clone(),
				#(#actor_fields),*
			}
		}?;

		let run_panic_handler: syn::Expr = panic_name
			.map(|handler| fallible_quote! { state.#handler(panic) })
			.unwrap_or(fallible_quote! { panic })?;

		let run_exit_handler: syn::Expr = exit_name
			.map(|handler| fallible_quote! { state.#handler()})
			.unwrap_or(fallible_quote! { () })?;

		let fun: ItemImpl = fallible_quote! {
			impl #actor_name {
				/// Creates a new Actor
				fn start(mut state: #data_name) -> shakespeare::ActorSpawn<#actor_name> {
					use ::shakespeare::{ActorSpawn, Channel, Context, catch_future, tokio_export as tokio};
					use ::std::sync::Arc;
					use tokio::{select, pin};
					use tokio::time::{sleep, Duration, Instant};

					const IDLE_TIMEOUT: Duration = Duration::from_secs(1);

					#(#queue_constructions)*
					let actor = Arc::new_cyclic(|weak| { #constructor });
					let stored_actor = Arc::clone(&actor);

					let mut context = Context::new(stored_actor);

					let event_loop = async move {
						let loop_lambda = async {
							let timeout_sleep = sleep(IDLE_TIMEOUT);
							pin!(timeout_sleep);
							loop {
								select! {
									#(#select_branches),*
									_ = &mut timeout_sleep, if context.sustains() => {
										if !context.sustains() {
											break;
										}
										else {
											timeout_sleep.as_mut().reset(Instant::now() + IDLE_TIMEOUT)
										}
									},
									else => { break; }
								};
							}
						};



						// SAFETY: The receive handles inside the branches are not safe to unwind
						// But they're consumed by the closure, so we can never see them
						// The senders might interact with a dead receiver though.
						// If we assume that a panic will not happen **during** an operation on the receiver,
						// then the control block will still be consistent at any point the sender looks at it
						// even if the receiver was destroyed
						let guarded_future = catch_future(loop_lambda);

						let result = guarded_future.await;

						match result {
							Ok(_) => { Ok(#run_exit_handler) },
							Err(panic) => Err(#run_panic_handler)
						}


					};

					let join_handle = tokio::task::spawn(event_loop);
					ActorSpawn::new(actor, join_handle)
				}
			}
		}?;

		Ok(SpawningFunction { fun })
	}
}

impl ToTokens for SpawningFunction {
	fn to_tokens(&self, tokens: &mut proc_macro2::TokenStream) {
		self.fun.to_tokens(tokens);
	}
}
