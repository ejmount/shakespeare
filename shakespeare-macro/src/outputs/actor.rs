use quote::ToTokens;
use syn::{Item, ItemFn, Result};

use crate::data::DataItem;
use crate::declarations::actor::ActorDecl;
use crate::macros::map_or_bail;
use crate::outputs::actor_struct::ActorStruct;
use crate::outputs::perfdispatch::PerfDispatch;
use crate::outputs::role::RoleOutput;
use crate::outputs::spawning_function::SpawningFunction;

#[derive(Debug)]
pub struct ActorOutput {
	data_item:         DataItem,
	actor_struct:      ActorStruct,
	spawning_function: SpawningFunction,
	panic_handler:     Option<ItemFn>,
	exit_handler:      Option<ItemFn>,
	performances:      Vec<PerfDispatch>,
	roles:             Vec<RoleOutput>,
	misc:              Vec<Item>,
}

impl ActorOutput {
	pub fn new(actor_node: ActorDecl) -> Result<ActorOutput> {
		let actor_struct = ActorStruct::new(&actor_node)?;

		let ActorDecl {
			actor_name,
			data_item,
			performances,
			roles,
			panic_handler,
			exit_handler,
			misc,
			..
		} = actor_node;

		let data_name = data_item.name();
		let panic_name = panic_handler.as_ref().map(|i| i.sig.ident.clone());
		let exit_name = exit_handler.as_ref().map(|i| i.sig.ident.clone());

		assert!(!performances.is_empty());
		let sf = SpawningFunction::new(
			&actor_name,
			&data_name,
			&performances,
			panic_name,
			exit_name,
		)?;

		let roles = map_or_bail!(roles, RoleOutput::new);

		let performances = map_or_bail!(&performances, |perf| PerfDispatch::new(
			perf,
			&actor_name,
			&data_name
		));

		let performances = performances.into_iter().flatten().collect();

		Ok(ActorOutput {
			data_item,
			performances,
			spawning_function: sf,
			roles,
			actor_struct,
			panic_handler,
			exit_handler,
			misc,
		})
	}
}

impl ToTokens for ActorOutput {
	fn to_tokens(&self, tokens: &mut proc_macro2::TokenStream) {
		self.actor_struct.to_tokens(tokens);
		self.data_item.to_tokens(tokens);
		self.spawning_function.to_tokens(tokens);
		self.panic_handler.to_tokens(tokens);
		self.exit_handler.to_tokens(tokens);
		for p in &self.performances {
			p.to_tokens(tokens);
		}
		for r in &self.roles {
			r.to_tokens(tokens);
		}
		for i in &self.misc {
			i.to_tokens(tokens);
		}
	}
}
