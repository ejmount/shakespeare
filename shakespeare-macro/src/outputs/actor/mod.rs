use quote::ToTokens;
use syn::{Item, ItemFn, Result};

mod actor_struct;
mod self_getter;
mod spawning_function;

use actor_struct::ActorStruct;
use self_getter::SelfGetter;
use spawning_function::SpawningFunction;

use super::performance::PerfDispatch;
use super::role::RoleOutput;
use crate::data::DataItem;
use crate::declarations::ActorDecl;
use crate::macros::map_or_bail;

#[derive(Debug)]
pub(crate) struct ActorOutput {
	data_item:         DataItem,
	actor_struct:      ActorStruct,
	getter:            SelfGetter,
	spawning_function: SpawningFunction,
	panic_handler:     Option<ItemFn>,
	exit_handler:      Option<ItemFn>,
	performances:      Vec<PerfDispatch>,
	roles:             Vec<RoleOutput>,
	misc:              Vec<Item>,
}

impl ActorOutput {
	pub(crate) fn new(actor_node: ActorDecl) -> Result<ActorOutput> {
		let actor_struct = ActorStruct::new(&actor_node)?;

		let ActorDecl {
			actor_name,
			data_item,
			performances,
			roles,
			panic_handler,
			exit_handler,
			misc,
			..
		} = actor_node;

		let data_name = data_item.name();
		let panic_name = panic_handler.as_ref().map(|i| i.sig.ident.clone());
		let exit_name = exit_handler.as_ref().map(|i| i.sig.ident.clone());

		let getter = SelfGetter::new(&actor_name)?;

		assert!(!performances.is_empty());
		let sf = SpawningFunction::new(
			&actor_name,
			&data_name,
			&performances,
			panic_name,
			exit_name,
		)?;

		let roles = map_or_bail!(roles, RoleOutput::new);

		let performances = map_or_bail!(&performances, |perf| PerfDispatch::new(
			perf,
			&actor_name,
			&data_name
		));

		let performances = performances.into_iter().flatten().collect();

		Ok(ActorOutput {
			data_item,
			performances,
			getter,
			spawning_function: sf,
			roles,
			actor_struct,
			panic_handler,
			exit_handler,
			misc,
		})
	}
}

impl ToTokens for ActorOutput {
	fn to_tokens(&self, tokens: &mut proc_macro2::TokenStream) {
		self.actor_struct.to_tokens(tokens);
		self.data_item.to_tokens(tokens);
		self.getter.to_tokens(tokens);
		self.spawning_function.to_tokens(tokens);
		self.panic_handler.to_tokens(tokens);
		self.exit_handler.to_tokens(tokens);
		for p in &self.performances {
			p.to_tokens(tokens);
		}
		for r in &self.roles {
			r.to_tokens(tokens);
		}
		for i in &self.misc {
			i.to_tokens(tokens);
		}
	}
}
