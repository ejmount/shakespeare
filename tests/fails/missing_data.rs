use shakespeare::actor;

#[actor]
mod actor {
	fn foo() {}
}

fn main() {}
