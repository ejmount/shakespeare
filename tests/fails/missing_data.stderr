error: Actor declaration must contain one struct, enum or union
 --> tests/fails/missing_data.rs:4:1
  |
4 | / mod actor {
5 | |     fn foo() {}
6 | | }
  | |_^
